\documentclass[a4paper,10pt]{article}
\usepackage[dvips]{color,graphicx}
\usepackage[dvips, bookmarks, colorlinks=false]{hyperref}
\usepackage[all]{xy}
\addtolength{\textheight}{7cm}
\addtolength{\textwidth}{3.5cm}
\addtolength{\hoffset}{-3cm}
\addtolength{\voffset}{-3cm}

\setlength{\parskip}{3ex plus 0.5ex minus 0.2ex}

%opening
\title{Paper}
\author{Yu Huang}

\begin{document}

\maketitle

\begin{abstract}

\end{abstract}

\section{Introduction}
Arabidopsis thaliana is a highly selfing species, with a very small portion resulting from outcrossing. Hartl et al\cite{Hartl2007} put the outcrossing rate to be around 1\%.

Given the polymorphism data of 149 SNPs in 4664 strains, we are able to estimate the outcrossing rate.

\section{Data}
The polymorphism data is collected by Justin Borevitz lab from University of Chicago. The affymetrix genotyping array is used. The 149 SNPs were picked according to the de-novo sequencing data\cite{Nordborg2005}. Figure~\ref{f1} shows the spacing of 149 SNPs. The probes are 25-mers with the changing middle base to detect two alleles. The 4664 strains were collected from around the world (Figure~(?) ).

\begin{figure}
\includegraphics[width=1\textwidth]{figures/snp_locations_on_chr.eps}
\caption{}\label{f1}
\end{figure}


\section{Genotyping Error Rate}
The 96 strains from \cite{Nordborg2005} were genotyped again, which offered an opportunity to have a rough estimate of genotyping error rate by comparing the data from two different experiments. Figure~ is an overview of all the differences.

The optimistic error rate (not counting the NAs) is 121/(12408+121) = 0.96\%. Taking the NAs into account, the error rate would be (121+893+804)/( 12408.0+121+893+804) = 12.78\%.


\section{Bogus Heterozygous Calls(column-wise)}
Figure~\ref{f2} shows the distribution of heterozygous calls in a Strain by SNP matrix. Apparently, there're columns with a number of unreasonable heterozygous calls.

\begin{figure}
\includegraphics[width=1\textwidth]{figures/justin_data_heterozygous.eps}
\caption{}\label{f2}
\end{figure}

A simple statistical model is used to quantitively tell how unreasonable each column is.

\subsection{model for snp locus}
\begin{equation}
P({SNP}_j|Strain\quad heterozygous\quad info) = \prod_{i=1}^{N} p_i^{a_i^j} (1-p_i)^{1-a_i^j}
\end{equation}

$p_i$ is probability that one strain has heterozygous call.

$a_i^j$ is indicator whether ${SNP}_j$ is homozygous($=1$) or not($=0$) for ${Strain}_i$.

Figure~\ref{f3} is the histogram of the probability for each SNP locus after taking logarithm.

\begin{figure}
\includegraphics[width=1\textwidth]{figures/justin_data_y_SNP_locus_log_prob.eps}
\caption{}\label{f3}
\end{figure}

4 SNP loci with log probability $\leq -0.5$ were removed. These four loci show long stretch of heterozygous calls among lots of strains in Figure~\ref{f2}.

\section{NA Filtering}
Figure ~\ref{f10} is the Strain by SNP matrix after being sorted. A bunch of NA-rich strains are shown on top of the figure.

\begin{figure}
\includegraphics[width=1\textwidth]{figures/justin_data_y_b.sorted.eps}
\caption{0 is NA. 1-4 is ACGT. 5-10 is heterozygous call}\label{f10}
\end{figure}


Figure~\ref{f4} is a histogram of NA percentage in all SNPs. 3 SNPs with $\geq 40\%$ NAs were removed. Figure~\ref{f5} is a histogram of NA percentage in all strains. 254 strains with $\geq 40\%$ NAs were removed.

\begin{figure}
\includegraphics[width=1\textwidth]{figures/justin_data_y_SNP_NA_perc.eps}
\caption{}\label{f4}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/justin_data_y_strain_NA_perc.eps}
\caption{}\label{f5}
\end{figure}

\section{Remove Identity Strains}
Due to the highly inbreeding nature of Arabidopsis, there're lots of identity strains in the samples collected from nature. In Figure~\ref{f10}, blocks of identity strains are very obvious. In the detection of identity strains, NA is regarded to be identical to any genotyping call. Because different strains have NAs in different SNP loci, the identity relation is not transitive. For example, A is identical to B and B is identical to C but A might not be identical to C. A graph is constructed with strains as nodes and two nodes are connected if these two strains are deemed as identical. An example graph is shown in Figure ~(?). A greedy algorithm is used to remove nodes with highest degree until no edges left.

In the end, 2575 strains were removed.

\section{Detecting Recombinant Inbred Line(RIL)}
Figure~\ref{f6} is a diagram showing the production of recombinant imbred lines by selfing.

use Dynamic programming to identify RILs with minimum number of recombinations. Regardless of how many selfing generations, the algorithm tells you whether one strain results from the outcrossing of two strains.

This step gives out results flooded with false trios. The 1st kind of false trio is that one parent is very close to the child, the other is a little farther (but still with substantial similarity) and just help to save some few incompatibilites. The 2nd kind is that one parent and child are close. The other parent is kind of irrelevant and saves the incompatibilites by being 'NA'.

\section{Estimate the number of selfing generations since the last outcrossing event}
Figure~\ref{f6} is a diagram showing the production of recombinant imbred lines by selfing. F1 would be 0 selfing generation away from the last outcrossing event as it's the direct result of outcrossing. F2 is 1 selfing generation away from the last outcrossing event.

\begin{figure}
\includegraphics[width=1\textwidth]{figures/selfing_diagram.eps}
\caption{from \cite{Broman2005}}\label{f6}
\end{figure}

The following model is from \cite{Enjalbert2000}.

\subsection{model to estimate the number of selfing generations since the last outcrossing event}
\begin{equation}
D_i = 1 - \sum_{j=1}^k p_{ij}^2
\end{equation}


$p_{ij}$ is probability of allele j of SNP i. k is the number of alleles.

$D_i$ is probability of SNP i being heterozygous.


\begin{equation}
likelihood(S_n) = \prod_{i=1}^L {(\frac{D_i}{2^n})}^{a_i} {(1-\frac{D_i}{2^n})}^{1-a_i}
\end{equation}

$S_n$ is the strain which is n selfing generations away from the last outcrossing event.

$L$ is the number of loci.

$a_i$ is indicator whether SNP i is heterozygous($=1$) or not($=0$).

$\hat{n}$ is the MLE.

This model assumes the independence among the SNP loci, which is not true in general due to LD. According the recent Kim et al.(2007) paper (in preparation), LD decays within 10kb on average. In our data, the spacing between neighboring SNP loci is generally huge (see Figure~\ref{f7}). If sorted, the list of gaps between loci looks like 1713, 4520, 13250, 49053, ..., 4482219, 6334701. So only the first two gaps are within 10kb frame. So statistical independence could be well assumed.

\begin{figure}
\includegraphics[width=1\textwidth]{figures/snp_locus_gap_hist.eps}
\caption{}\label{f7}
\end{figure}

Figure~\ref{f8} is the result. The spike in 6 is probably spurious estimates.

\begin{figure}
\includegraphics[width=1\textwidth]{figures/justin_data_y_b_filtered.estimate.selfing.generation.hist.eps}
\caption{}\label{f8}
\end{figure}

Figure~\ref{f9} is an example showing one strain with 0 selfing generations from the last outcrossing event.

\begin{figure}
\includegraphics[width=1\textwidth]{figures/trio1.eps}
\caption{}\label{f9}
\end{figure}

\section{Estimate selfing rate}
Here gives a summary of the methods tried so far. the goal is to see whether there're major difference among different methods. the quick answer is no.

\subsection{Jarne2006~\cite{Jarne2006}}
simplest, single-locus, $F_{IS}$-based. $F_{IS} = 1-H_{obs}/H_e$, where $H_e$ is the expected heterozygosity, $H_e = 2pq$ and $H_{obs}$ is the observed heterozygosity. $s = 2F_{IS}/(1+F_{IS})$.

\subsection{Robertson1984~\cite{Robertson1984}}
single-locus, $F_{IS}$-based. not significantly different from Jarne2006. estimate is slightly higher than Jarne2006's. in several cases, it gives estimate slightly over 1.

\subsection{Weir1984~\cite{Weir1984}}
single-locus, multi-loci, $F_{IS},F_{ST}$-based. The single locus estimate gives similar result to Jarne2006 or Nordborg1997. But the multi-locus gives significantly lower estimate.

\subsection{Nordborg1997~\cite{Nordborg1997}}
single-locus, colascent-based. This estimate gives value equal to Weir1984's single-locus one.

\subsection{David2007~\cite{David2007}}
it's based on identity disequilibrium, which is an excess of heterozygotes or homozygotes across multiple loci. this method is most computing intensive and shows greatest difference.

\subsection{todo}
\begin{enumerate}
 \item Enjalbert2000~\cite{Enjalbert2000} method
 \item Simulate data with different selfing rate and apply different algorithms to see how effective different methods are.
\end{enumerate}

How to define a population remains a question. Gao2007~\cite{Gao2007}, Pritchard2000~\cite{Pritchard2000}, Guillot2005~\cite{Guillot2005} and Francois2006~\cite{Francois2006} have interesting solutions. But before figuring out how to define populations, we just tried a simple method.

\subsection{Selfing rates across globe}
Each population is formed via connecting close sites. There's a distance (great circle distance) threshold to determine how close sites within a population should be. For each population, its data undergoes a few preprocessing procedures. 1. remove strains with $>=40\%$ NA. 2. remove snps with $>=40\%$ NA. 3. remove snps with too many heterozygous inconsistent with the strains' homozygous state (detailed in a previous section).

so far 5km (120 populations), 10km (96 populations), 25km (66 populations) have been tried. The figures below are shown in this order for each particular region (England, Europe Continent, North America, Sweden).

In figures below, selfing rates are estimated by Jarne2006. Other methods (except David2007's) show minor differences most of the time. we'll come to the differing parts later. Each population is labeled by selfing rate X 1000. 0 denotes not enough data to make inference (at least 5 samples/strains). The diameter of the circle corresponds to the size of the population.

\subsubsection{Standard deviation of these selfing rates}

Due to the fact that Jarne2006's estimate is single-locus based, the selfing rate that's put on the map is the average of all loci. For those low average selfing rates, the std (standard deviation) is also bigger, which means for most loci of that population, selfing rate estimate is pretty close to 1 and just a few with a lot of heterozygous calls which pulls down the average. However remember, bad snps with too many heterozygous inconsistent with the strains' homozygous state have already been eliminated. These ones are probably not technical errors.

\subsubsection{Negative selfing rate estimates}
In short, negative estimates are caused by the excess of heterozygosity compared to the Hardy Weinberg prediction ($2pq$). Supposedly, inbreeding would cause a deficiency of heterozygosity. Possibly due to sampling and extremely minor allele frequency, excess is observed. The identity-disequilibrium-based David2007's method doesn't suffer this problem.

It's hard to see those selfing rates without significantly magnification. I have high-resolution png figures.

\begin{figure}
\includegraphics[width=1\textwidth]{figures/s0829popid2ecotypeid_25_Eng__7_49_2_55_l3y1_pop_map.eps}
\caption{England}\label{f12}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/s0829popid2ecotypeid_10_Eng__7_49_2_55_l3y1_pop_map.eps}
\caption{England}\label{f11}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/s0829popid2ecotypeid_5_Eng__7_49_2_55_l3y1_pop_map.eps}
\caption{England}\label{f10}
\end{figure}


\begin{figure}
\includegraphics[width=1\textwidth]{figures/s0829popid2ecotypeid_25_EurCont__10_35_20_53_l3y1_pop_map.eps}
\caption{Europe Continent}\label{f15}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/s0829popid2ecotypeid_10_EurCont__10_35_20_53_l3y1_pop_map.eps}
\caption{Europe Continent}\label{f14}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/s0829popid2ecotypeid_5_EurCont__10_35_20_53_l3y1_pop_map.eps}
\caption{Europe Continent}\label{f13}
\end{figure}



\begin{figure}
\includegraphics[width=1\textwidth]{figures/s0829popid2ecotypeid_25_NorAm__88_38__72_45_l3y1_pop_map.eps}
\caption{North America}\label{f18}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/s0829popid2ecotypeid_10_NorAm__88_38__72_45_l3y1_pop_map.eps}
\caption{North America}\label{f17}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/s0829popid2ecotypeid_5_NorAm__88_38__72_45_l3y1_pop_map.eps}
\caption{North America}\label{f16}
\end{figure}


\begin{figure}
\includegraphics[width=1\textwidth]{figures/s0829popid2ecotypeid_25_Swe_10_52_20_65_l3y1_pop_map.eps}
\caption{Sweden}\label{f21}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/s0829popid2ecotypeid_10_Swe_10_52_20_65_l3y1_pop_map.eps}
\caption{Sweden}\label{f20}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/s0829popid2ecotypeid_5_Swe_10_52_20_65_l3y1_pop_map.eps}
\caption{Sweden}\label{f19}
\end{figure}

\subsection{identity strains across globe}
Here is to see how identity strain pairs are distributed across globe.

5002 strains with latitude and longitude. Same preprocessing procedures (removal for strains and snps with too many NAs and removal of bad highly heterozygous snp loci) applied. 4805 strains left.

In total, there're 426,913 identity pairs. First check how many different kinds of identity pairs (connected components if viewing them as a graph). (will add later).

281207(65.9\%) pairs are across-population for population model defined by 25km. In figure~\ref{f22}, Each population is denoted as a red dot with its diameter proportional to the number of identity pairs within that population. An edge is connected if there's identity relationship between two populations. The thickness of an edge corresponds to the number identity pairs between them.

\begin{figure}
\includegraphics[width=1\textwidth]{figures/popid2ecotypeid_25_identity.map_site_network.eps}
\caption{Population Identity Map}\label{f22}
\end{figure}


\bibliography{outcrossing}
\bibliographystyle{plain}

\end{document}
