% Template for PLoS
% Version 1.0 January 2009
%
% To compile to pdf, run:
% latex plos.template
% bibtex plos.template
% latex plos.template
% latex plos.template
% dvipdf plos.template

\documentclass[10pt]{article}

% amsmath package, useful for mathematical formulas
\usepackage{amsmath}
% amssymb package, useful for mathematical symbols
\usepackage{amssymb}

% graphicx package, useful for including eps and pdf graphics
% include graphics with the command \includegraphics
\usepackage{graphicx}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

\usepackage{color} 

% Use doublespacing - comment out for single spacing
%\usepackage{setspace} 
%\doublespacing


% Text layout
\topmargin 0.0cm
\oddsidemargin 0.5cm
\evensidemargin 0.5cm
\textwidth 16cm 
\textheight 21cm

% Bold the 'Figure #' in the caption and separate it with a period
% Captions will be left justified
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}

% Use the PLoS provided bibtex style
\bibliographystyle{plos2009}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother


% Leave date blank
\date{}

\pagestyle{myheadings}
%% ** EDIT HERE **


%% ** EDIT HERE **
%% PLEASE INCLUDE ALL MACROS BELOW

%% END MACROS SECTION

\begin{document}

% Title must be 150 characters or less
\begin{flushleft}
{\Large
\textbf{Enrichment Analysis of collective \textit{Arabidopsis thaliana} GWAS reveals ...}
}
% Insert Author names, affiliations and corresponding author email.
\\
Author1$^{1}$, 
Author2$^{2}$, 
Magnus Nordborg$^{3,\ast}$
\\
\bf{1} Author1 Dept/Program/Center, Institution Name, City, State, Country
\\
\bf{2} Author2 Dept/Program/Center, Institution Name, City, State, Country
\\
\bf{3} Author3 Dept/Program/Center, Institution Name, City, State, Country
\\
$\ast$ E-mail: Corresponding author@institute.edu
\end{flushleft}

% Please keep the abstract between 250 and 300 words
\section*{Abstract}
We also examined the representation of functional gene categories (gene ontologies) containing one or more associations among top GWAS results. highly suggestive loci for a variety of traits that did not meet genome-wide significant thresholds in prior analyses

% Please keep the Author Summary between 150 and 200 words
% Use first person. PLoS ONE authors please skip this step. 
% Author Summary not valid for PLoS ONE submissions.   
\section*{Author Summary}

\section*{Introduction}

association method comparison part:

phenotype whose association results show big overlap (by a statistics) among different association methods

	One figure shows an example.

phenotype whose association results show little overlap among different association methods

	Figure shows an example.
	
Table listing all phenotypes sorted in the order of the overlapping statistics.

discuss why? (look at how accurately the whole genotype could predict phenotype?)


Enrichment part:

The method corrects for the presence of linkage disequilibrium (LD) between SNPs, variable gene size, overlapping genes, and multiple nonindependent GO categories.

	Figure of how the enrichment ratio is calculated.

	Figure of how the pvalue of the enrichment ratio is derived.

classify phenotypes into those that the enrichment is near the top, in the middle, and no enrichment for corresponding candidate gene lists (if suitable lists exist).

look at enrichment of GO categories

1. new pathways significantly involved in certain category of phenotypes,

	table/figure shows an example

2. constitutive pathways across diverse phenotypes

	Figure shows an example



% Results and Discussion can be combined.
\section*{Results}



Optional parts:
0. predict phenotype from genotype to see which phenotype is good, which is not.

1. CMP (Kimmel's method) gives interval (like between two SNPs) results

2. weighted score sum, training using GO enrichment

3. overlapping of regions/peaks found. R-statistics (scan) or recombination rate or HMM to delimit the regions for each method

4. Signficance of difference in the SNPs identified by each method for each phenotype or phenotype class.

5. GO enrichment for genes identified in each relevant portion of the Venn diagram, for all tests, and for all genes found.

6. Are the TFs more-pleiotropic than random genes?

7. pleiotropic genes, pathways, groups of TFs

8. investigate the improvement of increasing the sample-size and gene coverage given by the imputed data


\subsection*{Subsection 1}

\subsection*{Subsection 2}

\section*{Discussion}

% You may title this section "Methods" or "Models". 
% "Models" is not a valid title for PLoS ONE authors. However, PLoS ONE
% authors may use "Analysis" 
\section*{Materials and Methods}

% Do NOT remove this, even if you are not including acknowledgments
\section*{Acknowledgments}


%\section*{References}
% The bibtex filename
\bibliography{template}

\section*{Figure Legends}
%\begin{figure}[!ht]
%\begin{center}
%%\includegraphics[width=4in]{figure_name.2.eps}
%\end{center}
%\caption{
%{\bf Bold the first sentence.}  Rest of figure 2  caption.  Caption 
%should be left justified, as specified by the options to the caption 
%package.
%}
%\label{Figure_label}
%\end{figure}


\section*{Tables}
%\begin{table}[!ht]
%\caption{
%\bf{Table title}}
%\begin{tabular}{|c|c|c|}
%table information
%\end{tabular}
%\begin{flushleft}Table caption
%\end{flushleft}
%\label{tab:label}
% \end{table}

\end{document}

