\documentclass[a4paper,10pt]{article}
\usepackage[pdftex]{color,graphicx}
\usepackage{ifpdf}
\ifpdf
\usepackage[pdftex, bookmarks, colorlinks, plainpages=false,pdfpagelabels,breaklinks]{hyperref}
\else
\usepackage[dvips, bookmarks, colorlinks, linkcolor=blue, urlcolor=blue, citecolor=blue, plainpages=false,pdfpagelabels,linktocpage]{hyperref}
\fi

\usepackage{longtable}
%\usepackage{afterpage}
\usepackage{float}
%\usepackage[all]{xy}
\addtolength{\textheight}{5cm}
\addtolength{\textwidth}{5cm}
\addtolength{\hoffset}{-3cm}
\addtolength{\voffset}{-3cm}

%opening
\title{the investigation of 149-snp peak data}
\author{Yu Huang}

\begin{document}

\maketitle

\begin{abstract}
look at the peak data for different calls
\end{abstract}

\tableofcontents


\section{cluster plots}
282802 out of 955388 snp calls (30\%) has raw peak(intensity) data. check their cluster plots.

For every snp genotyped, there're two alleles. Sequenom uses mass spectrometer to distinguish different alleles. It produces the intensity of peak1 and peak2, which corresponds to allele 1 and 2 respectively.

A cluster plot (name comes from the supplementary of the Wellcome Trust 2007 paper. \url{https://dl324b-1.cmb.usc.edu/research/ref/nobody2007}) is a graphical representation of the results of both the genotyping and genotype calling of a SNP. It is a scatter plot of the two peak intensities from the genotyping, with each point representing one snp calling in one ecotype/strain. Each point is colored according to how genotype calling algorithm decided to classify that particular snp in one ecotype. 4 different classifications are homozygote for allele 1 (red), homozygote for allele 2 (blue), heterozygote for two alleles (green), missing/NA (yellow).

Figure~\ref{f1}-\ref{f7} are cluster plots with all snps concatenated together.

Figure~\ref{flAtMSQTsnp2} and onwards are the cluster plots for one single snp (ordered by chromosome, position). the genotype calling algorithm might have different parameters for different snps.

One apparent problem is that same color dots appear on two off-diagonal sides. This can be seen in some single-snp cluster plots. (2007-10-30) Figure~\ref{f2} and \ref{f1} shows this issue is not as serious for allele 2 homozygote as for allele 1 homozygote. it looks like it doesn't happen at all to allele 2 homozygote calling (dots all above the diagonal line in Figure~\ref{f2}). this  raises the suspicion that the sequenom algorithm uses the diagonal line to separate data. this diagonal cut is also visible in Figure~\ref{f1}.

Three Possibilities:

1. some peak1/peak2 data was swapped in the database.

2. the sequenom genotype calling algorithm would be very lousy by calling some points from both sides of the heterozygous cloud as the same allele.

3. the correponding allele for peak1/peak2 is changing even for the same snp.


we need to check this to get a solid judgement.

but for some snps, like Figure~\ref{flAtMSQTsnp15}, \ref{flAtMSQTsnp54}, \ref{flAtMSQTsnp62}, \ref{flAtMSQTsnp97}, \ref{flAtMSQTsnp129}, \ref{flAtMSQTsnp132}, \ref{flAtMSQTsnp138}, \ref{flAtMSQTsnp173}, \ref{flAtMSQTsnp237}, \ref{flAtMSQTsnp263}, \ref{flAtMSQTsnp267}, \ref{flAtMSQTsnp304}, \ref{flAtMSQTsnp312}, \ref{flAtMSQTsnp334}, etc, the cluster plots don't look good. some of them might be good after we figure out the cause of same-color-both-side problem. they might correspond to snps with too many hets , or too many NAs.

(2007-10-30) one more thing is dots sitting on the X or Y axis. One of the peak1/peak2 is 0. is that real 0? given the noise in MS(mass spectrometry), it's hard to get absolute 0, isn't it? i'm no MS expert, willing to hear feedback.


\begin{figure}
\includegraphics[width=1\textwidth]{figures/cluster_plots_allele1.png}
\caption{homozygote for allele 1}\label{f1}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/cluster_plots_allele2.png}
\caption{homozygote for allele 2}\label{f2}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/cluster_plots_het.png}
\caption{heterozygote}\label{f3}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/cluster_plots_NA.png}
\caption{missing/NA}\label{f4}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/cluster_plots_allele1_het.png}
\caption{overlapping of homozygote for allele 1 and heterozygote}\label{f5}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/cluster_plots_allele2_het.png}
\caption{overlapping of homozygote for allele 2 and heterozygote}\label{f6}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/cluster_plots_allele1_allele2.png}
\caption{overlapping of homozygote for allele 1 and allele 2}\label{f7}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/cluster_plots_allele1_NA.png}
\caption{overlapping of homozygote for allele 1 and NA}\label{f8}
\end{figure}

\begin{figure}
\includegraphics[width=1\textwidth]{figures/cluster_plots_homo_and_hetero.png}
\caption{homozygote and heterozygote}\label{f9}
\end{figure}


\begin{figure}
\includegraphics[width=1\textwidth]{figures/cluster_plots_all.png}
\caption{all four different calls}\label{f10}
\end{figure}

\subsection{cluster plots for single snp}

\input{tables_figures.tex}

\end{document}
